src/esperanza/finalizationstate.cpp

```diff
@@ -168,13 +169,12 @@ void FinalizationState::IncrementDynasty() {
     m_current_dynasty += 1;
     m_prev_dyn_deposits = m_cur_dyn_deposits;
     m_cur_dyn_deposits += GetDynastyDelta(m_current_dynasty);
-    m_dynasty_start_epoch[m_current_dynasty] = m_current_epoch;
+    m_dynasty_start_epoch[m_current_dynasty] = m_current_epoch + 1;
```

Comment:
It's quite strange that `IncrementDynasty` is invoked in `InitializeEpoch` _before_ `m_current_epoch` is incremented.

Reply:
@Nizametdinov I agree that it looks weird and thinking about this again, it makes sense to invoke `IncrementDynasty ` after the `InitializeEpoch`. I had that in mind but didn't do it to keep the diff as small as possible.

Reply:
I think that these are two different problems, what @Nizametdinov is pointing could be considered as a refactor. I would say that it's ok to apply the minimum changes to fix the issue.

By the way, does the fixed issue have an issue number to refer to?

Reply:
@castarco no, I didn't create the issue for this one as I found out about the problem during another investigation I am doing. I think we can create the issue first if we don't intend to work on it right away or it requires some input. Creating a simple placeholder instead of the adequately described issue is also not good as people won't able to follow it. And I think creating the GitHub PR is also the same as creating the issue with an already ready example to reproduce it :) And if we agree that it is actually a problem but we don't like the solution, we can convert it to the issue any time

Reply:
I thought that moving `IncrementDynasty` in `InitializeEpoch` few lines lower is a quite simple change which makes `m_current_epoch + 1` unnecessary here. If it's not that straightforward it would be better to do it in a separate PR.

Reply:
@Nizametdinov before we increment the epoch, we also re-compute `m_cur_dynasty_deposits` and `m_prev_dynasty_deposits` and it seems there is an issue there which I am investigating. I'd like to keep it as it is and once I write tests to (dis)proof it, will change the order of how we invoke `InitializeEpoch` and `IncrementDynasty`.